core/kernel/transfer_list.c

```diff
@@ -0,0 +1,580 @@
+		 * ensure that the increased size doesn't exceed
+		 * the max size of TL
+		 */
+		mov_dis = new_ev - old_ev;
+		if (ROUNDUP_OVERFLOW(mov_dis,
+				     TL_ALIGNMENT_FROM_ORDER(tl->alignment),
+				     &mov_dis) ||
+		    tl->size + mov_dis > tl->max_size) {
+			return false;
+		}
+		ru_new_ev = old_ev + mov_dis;
+		memmove((void *)ru_new_ev, (void *)old_ev, tl_old_ev - old_ev);
+		tl->size += mov_dis;
+		gap = ru_new_ev - new_ev;
+	} else {
+		gap = old_ev - new_ev;
+	}
+
+	if (gap >= sizeof(*dummy_te)) {
+		/* create a dummy TE to fill up the gap */
+		dummy_te = (struct transfer_list_entry *)new_ev;
+		dummy_te->tag_id = TL_TAG_EMPTY;
+		dummy_te->reserved0 = 0;
+		dummy_te->hdr_size = sizeof(*dummy_te);
+		dummy_te->data_size = gap - sizeof(*dummy_te);
+	}
+
+	te->data_size = new_data_size;
+
+	transfer_list_update_checksum(tl);
+	return true;
+}
+
+/*******************************************************************************
+ * Remove a specified transfer entry from a transfer list.
+ * @tl: Pointer to the transfer list.
+ * @te: Pointer to the transfer entry.
+ * Return true on success or false on error.
+ ******************************************************************************/
+bool transfer_list_rem(struct transfer_list_header *tl,
+		       struct transfer_list_entry *te)
+{
+	if (!tl || !te || (vaddr_t)te > (vaddr_t)tl + tl->size)
+		return false;
+
+	te->tag_id = TL_TAG_EMPTY;
+	te->reserved0 = 0;
+	transfer_list_update_checksum(tl);
+	return true;
+}
+
+/*******************************************************************************
+ * Add a new transfer entry into a transfer list.
+ * Compliant with 2.4.3 of Firmware Handoff specification (v0.9).
+ * @tl: Pointer to the transfer list.
+ * @tag_id: Tag ID for the new transfer entry.
+ * @data_size: Data size of the new transfer entry.
+ * @data: Pointer to the data for the new transfer entry.
+ * Return pointer to the added transfer entry or NULL on error.
+ ******************************************************************************/
+struct transfer_list_entry *transfer_list_add(struct transfer_list_header *tl,
+					      uint16_t tag_id,
+					      uint32_t data_size,
+					      const void *data)
+{
+	vaddr_t max_tl_ev = 0;
+	vaddr_t tl_ev = 0;
+	vaddr_t ev = 0;
+	struct transfer_list_entry *te = NULL;
+	uint8_t *te_data = NULL;
+	size_t sz = 0;
+
+	if (!tl)
+		return NULL;
+
+	max_tl_ev = (vaddr_t)tl + tl->max_size;
+	tl_ev = (vaddr_t)tl + tl->size;
+	ev = tl_ev;
+
+	/*
+	 * skip the step 1 (optional step)
+	 * new TE will be added into the tail
+	 */
+	if (ADD_OVERFLOW(sizeof(*te), data_size, &sz) ||
+	    ADD_OVERFLOW(ev, sz, &ev) ||
+	    ROUNDUP_OVERFLOW(ev, TRANSFER_LIST_GRANULE, &ev) ||
+	    ev > max_tl_ev) {
+		return NULL;
+	}
+
+	te = (struct transfer_list_entry *)tl_ev;
+	*te = (struct transfer_list_entry){
+		.tag_id = tag_id,
+		.hdr_size = sizeof(*te),
+		.data_size = data_size,
+	};
+
+	tl->size += ev - tl_ev;
```

Comment:
I think `tl->size` should be updated only once `data` processing below succeeds

Reply:
No. This function allows passing in NULL TE data pointer.
In this case, we still create the TE with the required size but without copying the data. The data region is just for placeholder.
The scenario for this is to add a 'gap' with an 'empty' tag to fulfill the alignment requirement, see the caller function `transfer_list_add_with_align()` for reference.